### ISSUE TITLE
Cannot launch VS Code debugger with STM32

### ISSUE THREAD

**Issue by ademyankov** [Azure-Samples/iot-middleware-freertos-samples#120]
Well, basically nothing works out of the box or by just simply following the [instructions](https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/main/demos/projects/ST/b-l475e-iot01a/VSCodeDebug.md):

The first problem is that the PowerShell script simply doesn't work:

![image](https://user-images.githubusercontent.com/13845895/141350514-4890b14d-02f7-4946-94b0-35e3eed2cb99.png)

However, I was able to execute the script manually from the CMD and it successfully fetched the RTOS.

If I skip it I hit the problem with openocd:

![image](https://user-images.githubusercontent.com/13845895/141352798-72d6ce0d-3a48-42e3-8bc9-d67f67e3d966.png)

I tried to add -s command to openocd:
![image](https://user-images.githubusercontent.com/13845895/141358729-853293cc-9c24-404e-b4b5-d0d2eeab41e8.png)

I can launch it manually from the power shell:

![image](https://user-images.githubusercontent.com/13845895/141358551-ce741907-d3cb-4dd4-a533-c772266a996f.png)

OpenOCD folder looks like this:
![image](https://user-images.githubusercontent.com/13845895/141358609-c49d74c4-3086-43e6-b625-67e5445dc835.png)

PATH variable points to C:\ST\OpenOCD-20210729-0.11.0

I have noticed a weird thing...

![image](https://user-images.githubusercontent.com/13845895/141381854-5d7ad232-eed3-4944-b847-5ee73fd926b7.png)

why does it add '-s' and the root folder of the IoT project for the openocd cfg search path? 

I couldn't find any JSON settings explicitly doing this.

After I copied folders from the openocd into the root of the project at least the debugger starts.

**Since you guys went through this hassle can you see or possibly identify where the problem might be with all those things?**

**Comment by danewalton** [comment]
Hey @ademyankov thanks for moving this over.

At least for the freertos download, [we have a PR here which should fix that](https://github.com/Azure-Samples/iot-middleware-freertos-samples/pull/118). Looks like your default integrated terminal is `cmd` so it will need the explicit powershell invocation to run it properly.

I have a feeling the `cmd` vs `powershell` idea is also what is causing the other parts to fail which is why the terminal with the blue background you posted (powershell) is working, while running it from VSCode does not. I can look into making sure it works if `cmd` is the default terminal.

**Comment by ademyankov** [comment]
@danewalton thanks, man, good to know. Actually, I don’t give a d* about cmd :) if I can make powershell as a default terminal and everything will work I would easily go with that.

Can you share the trick to make it happen?

**Comment by danewalton** [comment]
Oh wonderful that helps then :)

If you go to this setting in VSCode you should be able to change it to run Powershell.

![image](https://user-images.githubusercontent.com/36710865/141510008-9772bf38-d36c-4cff-bf43-a4e8a3d8639f.png)

Actually change this one too since this says specifically for "tasks"

![image](https://user-images.githubusercontent.com/36710865/141510232-00a34948-1949-4c93-a36f-1a7f8d61b4f6.png)

Should look like this

![image](https://user-images.githubusercontent.com/36710865/141510328-dd2d24e7-1a75-4ddc-b530-5220b653316c.png)

**Comment by ademyankov** [comment]
Ok, I got a bit further. Now, it uses PowerShell, it fetches FreeRTOS, builds the code but still has issues with openocd.

![image](https://user-images.githubusercontent.com/13845895/141597960-0418cb24-208d-4f7f-9928-82a4355e3fcd.png)

**Comment by ademyankov** [comment]
Just wanted to let you know if I copy all the stuff from openocd: board, interface, and target into the root of the project then the debugger starts and all works as expected.

![image](https://user-images.githubusercontent.com/13845895/141598335-9d28c27d-5b4d-408b-aa07-5a88048993fc.png)

**Comment by danewalton** [comment]
Okay did a bit of digging. The error message you're getting is

> Error: Can't find board/stm32l4discovery.cfg

[On this page of the docs for OpenOCD](https://elinux.org/OpenOCD_Ubuntu_Package) it says the following

> The package installs config files to /usr/share/openocd/scripts, but OpenOCD looks for config files in /usr/local/share/openocd/scripts. Fix that now by creating a symbolic link in /usr/local/share to /usr/share/openocd. In the terminal window, type:
> ```bash
> cd /usr/local/share
> sudo ln -s /usr/share/openocd .
> ```
> OpenOCD will now be able to find its config files. For more information on how OpenOCD searches for config files, see OpenOCD Config File Paths.

OpenOCD first looks in the "current directory" to find the config file, so when you say you copy the file to the top level, it makes sense it would find it. 

I don't think you're on Linux but it's possible you might be having a similar issue?

**Comment by ademyankov** [comment]
Yes, @danewalton, I am not on Linux. I am on Windows 10.

**Comment by danewalton** [comment]
The fact that putting the board configuration in the top level allows it to work has me believing some path variable isn't set up correctly. Are you able to see the attached screenshot and the link provided to make sure the OpenOCD environment is set up correctly?

![image](https://user-images.githubusercontent.com/36710865/142059318-0c05a986-8c70-4612-9520-01b34fb0c694.png)


https://elinux.org/Running_OpenOCD_on_Windows

**Comment by ademyankov** [comment]
Yes, I have seen this page... and my windows env var points to openocd folder where all those directories are. But as I pointed it out before, somewhere, some settings changes the search path for openocd tool with '-s' switch.

![image](https://user-images.githubusercontent.com/13845895/142067235-4ee74ec4-87f6-4ed8-8ce3-36722f173102.png)

As you can see it points to the root of the project.

I tried manually add that switch into the task.json to point it to the right directory:

![image](https://user-images.githubusercontent.com/13845895/142067414-b300e7fc-7a8f-4a2c-8329-28d586db0afd.png)

But it did not help at all.

I will dig further and if I find a solution I will definetely add a comment here.

**Comment by ademyankov** [comment]
Oh, BTW, @danewalton, I have a question.

When the debuger is run it downloads and builds FreeRTOS repo via `.github/scripts/fetch_freertos.ps1` script into **lib/FreeRTOS** folder and then firmware binary is built using this version of FreeRTOS.

The question is what is the reason for doing so? Because, if the project is built without debugging then the FreeRTOS is fetched via cmake script into **b-l475e-iot01a/_deps/freertos-src** folder. So, basically the project already has FreeRTOS fetched and built.
